File: [0200][Medium][Number_of_Islands]/Number_of_Islands_4.java
// leetcode time     cost : 3 ms
// leetcode memory   cost : 42 MB 
// Time  Complexity: O(M*N)
// Space Complexity: O(M*N)
// solution 1, DFS,
// https://leetcode.com/problems/number-of-islands/discuss/56349/7-lines-Python-~14-lines-Java
public class Solution {
  char[][] g;
  public int numIslands(char[][] grid) {
      int islands = 0;
      g = grid;
      for (int i=0; i<g.length; i++)
          for (int j=0; j<g[i].length; j++)
              islands += sink(i, j);
      return islands;
  }
  int sink(int i, int j) {
      if (i < 0 || i == g.length || j < 0 || j == g[i].length || g[i][j] == '0')
          return 0;
      g[i][j] = '0';
      sink(i+1, j); sink(i-1, j); sink(i, j+1); sink(i, j-1);
      return 1;
  }
}

public class Number_of_Islands_4 {
    public static void main(String args[]) {
        char[][] grid = {{'1','1','1','1','0'},{'1','1','0','1','0'},{'1','1','0','0','0'},{'0','0','0','0','0'}}; // expect is 1,
        Solution Solution_obj = new Solution();
        int result = Solution_obj.numIslands(grid);
        System.out.println("In java code, 1st result is :" + result);
    }
}

File: [0200][Medium][Number_of_Islands]/Number_of_Islands_4.py
# leetcode time     cost : 192 ms
# leetcode memory   cost : 15.5 MB 
# Time  Complexity: O(M*N)
# Space Complexity: O(M*N)
# solution 3. Union Find solution, with both path compression and union by rank
class UnionFind(object):
    def __init__(self, grid):
        m, n = len(grid), len(grid[0])
        self.count = 0
        self.parent = [-1] * (m*n)
        self.rank = [0] * (m*n)
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    self.parent[i*n + j] = i*n + j
                    self.count += 1

    def find(self, i):
        if self.parent[i] != i:
            self.parent[i] = self.find(self.parent[i])
        return self.parent[i]

    def union(self, x, y):
        rootx = self.find(x)
        rooty = self.find(y)
        if rootx != rooty:
            if self.rank[rootx] > self.rank[rooty]:
                self.parent[rooty] = rootx
            elif self.rank[rootx] < self.rank[rooty]:
                self.parent[rootx] = rooty
            else:
                self.parent[rooty] = rootx
                self.rank[rootx] += 1
            self.count -= 1

class Solution(object):
    def is_valid(self, grid, r, c):
        m, n = len(grid), len(grid[0])
        if r < 0 or c < 0 or r >= m or c >= n:
            return False
        return True

    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid or not grid[0]:
            return 0

        uf = UnionFind(grid)

        directions = [(0,1), (0,-1), (-1,0), (1,0)]
        m, n = len(grid), len(grid[0])
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    for d in directions:
                        nr, nc = i + d[0], j + d[1]
                        if self.is_valid(grid, nr, nc) and grid[nr][nc] == '1':
                            uf.union(i*n+j, nr*n+nc)
        return uf.count

def main():
    grid = [["1","1","1","1","0"],["1","1","0","1","0"],["1","1","0","0","0"],["0","0","0","0","0"]]  # expect is 1,
    obj = Solution()
    result = obj.numIslands(grid)       
    print("return result is ",result)
    
if __name__ =='__main__':
    main() 